modules/gateway/gateway.go

```diff
@@ -159,18 +145,26 @@ func New(addr string, persistDir string) (g *Gateway, err error) {
 	g.log.Println("INFO: gateway created, started logging")
 
 	// Forward the RPC port, if possible.
-	go g.forwardPort(port)
+	g.threads.Add()
+	go func() {
+		defer g.threads.Done()
+		g.forwardPort(port)
+	}()
 
 	// Learn our external IP.
-	go g.learnHostname(port)
+	g.threads.Add()
+	go func() {
+		defer g.threads.Done()
```

Comment:
The advantage of doing the `Add` outside the goroutine is that you do not need to check the return value, because you know that the gateway cannot be closed inside the `New` function.